test/QueryPlannerTest.cpp

```diff
@@ -1600,18 +1603,542 @@ TEST(QueryPlanner, PathSearchWrongArgumentAlgorithm) {
+  h::expect(
+      "PREFIX spatialSearch: <https://qlever.cs.uni-freiburg.de/spatialSearch/>"
+      "SELECT * WHERE {"
+      "?x <p> ?y."
+      "SERVICE spatialSearch: {"
+      "_:config spatialSearch:algorithm spatialSearch:s2 ;"
+      "spatialSearch:left ?y ;"
+      "spatialSearch:right ?b ;"
+      "spatialSearch:maxDistance 100 ;"
+      "spatialSearch:nearestNeighbors 2 ;"
+      "spatialSearch:bindDistance ?dist ."
+      "{ ?a <p> ?b } }}",
+      h::SpatialJoin(100, 2, V{"?y"}, V{"?b"}, V{"?dist"}, emptyPayload, S2,
+                     scan("?x", "<p>", "?y"), scan("?a", "<p>", "?b")));
+  h::expect(
+      "PREFIX spatialSearch: <https://qlever.cs.uni-freiburg.de/spatialSearch/>"
+      "SELECT * WHERE {"
+      "?x <p> ?y."
+      "SERVICE spatialSearch: {"
+      "_:config spatialSearch:algorithm spatialSearch:s2 ;"
+      "spatialSearch:right ?b ;"
+      "spatialSearch:nearestNeighbors 5 . "
+      "_:config spatialSearch:left ?y ."
+      "{ ?a <p> ?b } }}",
+      h::SpatialJoin(-1, 5, V{"?y"}, V{"?b"}, std::nullopt, emptyPayload, S2,
+                     scan("?x", "<p>", "?y"), scan("?a", "<p>", "?b")));
+}
+
+TEST(QueryPlanner, SpatialJoinServicePayloadVars) {
+  // Test the <payload> option which allows selecting columns from the graph
+  // pattern inside the service.
+
+  auto scan = h::IndexScanFromStrings;
+  using V = Variable;
+  auto S2 = SpatialJoinAlgorithm::S2_GEOMETRY;
+
+  h::expect(
+      "PREFIX spatialSearch: <https://qlever.cs.uni-freiburg.de/spatialSearch/>"
+      "SELECT * WHERE {"
       "?x <p> ?y."
-      "?a <p> ?b."
-      "?y <max-distance-in-meters:1> ?b }",
-      h::SpatialJoin(1, -1, scan("?x", "<p>", "?y"), scan("?a", "<p>", "?b")));
+      "SERVICE spatialSearch: {"
+      "_:config spatialSearch:algorithm spatialSearch:s2 ;"
+      "spatialSearch:right ?b ;"
+      "spatialSearch:bindDistance ?dist ;"
+      "spatialSearch:nearestNeighbors 5 . "
+      "_:config spatialSearch:left ?y ."
+      "_:config spatialSearch:payload ?a ."
+      "{ ?a <p> ?b } }}",
+      h::SpatialJoin(-1, 5, V{"?y"}, V{"?b"}, V{"?dist"},
+                     std::vector<V>{V{"?a"}}, S2, scan("?x", "<p>", "?y"),
+                     scan("?a", "<p>", "?b")));
+  h::expect(
+      "PREFIX spatialSearch: <https://qlever.cs.uni-freiburg.de/spatialSearch/>"
+      "SELECT * WHERE {"
+      "?x <p> ?y."
+      "SERVICE spatialSearch: {"
+      "_:config spatialSearch:algorithm spatialSearch:s2 ;"
+      "spatialSearch:right ?b ;"
+      "spatialSearch:bindDistance ?dist ;"
+      "spatialSearch:nearestNeighbors 5 . "
+      "_:config spatialSearch:left ?y ."
+      "_:config spatialSearch:payload ?a , ?a2 ."
+      "{ ?a <p> ?a2 . ?a2 <p> ?b } }}",
+      h::SpatialJoin(
+          -1, 5, V{"?y"}, V{"?b"}, V{"?dist"},
+          std::vector<V>{V{"?a"}, V{"?a2"}}, S2, scan("?x", "<p>", "?y"),
+          h::Join(scan("?a", "<p>", "?a2"), scan("?a2", "<p>", "?b"))));
```

Comment:
Who tests/handles the case of `spatialSerach: payload ?a, ?a, ?a.` (the same payload multiple times?) Does that silently deduplicate, throw an exception, or crash?

Reply:
The test is in the `TEST_P(SpatialJoinVarColParamTest, payloadVariables)` test case. Multiple occurrences are silently deduplicated by the `computeVariableToColumnMap()` and `prepareJoin()` functions (they use the `getVarColMapPayloadVars()` function to be more specific). I have updated the `getResultWidth()` function to match this behavior. The same is true if the user specifies the right join variable in the payload. If variables from the left table are included an explanatory error message is thrown.